miner/tx_pool.go

```diff
@@ -0,0 +1,109 @@
+package miner
+
+import (
+	"errors"
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/pending_txs"
+	"github.com/spacemeshos/go-spacemesh/rand"
+	"sync"
+)
+
+type TxMempool struct {
+	txs      map[types.TransactionId]*types.Transaction
+	accounts map[types.Address]*pending_txs.AccountPendingTxs
+	mu       sync.RWMutex
+}
+
+func NewTxMemPool() *TxMempool {
+	return &TxMempool{
+		txs:      make(map[types.TransactionId]*types.Transaction),
+		accounts: make(map[types.Address]*pending_txs.AccountPendingTxs),
+	}
+}
+
+func (t *TxMempool) Get(id types.TransactionId) (types.Transaction, error) {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	if tx, found := t.txs[id]; found {
+		return *tx, nil
+	}
+	return types.Transaction{}, errors.New("transaction not found in mempool")
+}
+
+func (t *TxMempool) GetTxsForBlock(numOfTxs int, getState func(addr types.Address) (nonce, balance uint64, err error)) ([]types.TransactionId, error) {
+	var txIds []types.TransactionId
+	t.mu.RLock()
+	for addr, account := range t.accounts {
+		nonce, balance, err := getState(addr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get state for addr %s: %v", addr.Short(), err)
+		}
+		accountTxIds, _, _ := account.ValidTxs(nonce, balance)
+		txIds = append(txIds, accountTxIds...)
+	}
+	t.mu.RUnlock()
+
+	if len(txIds) <= numOfTxs {
+		return txIds, nil
+	}
+
+	var ret []types.TransactionId
+	for idx := range getRandIdxs(numOfTxs, len(txIds)) {
+		//noinspection GoNilness
+		ret = append(ret, txIds[idx])
+	}
+	return ret, nil
+}
+
+func getRandIdxs(numOfTxs, spaceSize int) map[uint64]struct{} {
+	idxs := make(map[uint64]struct{})
+	for len(idxs) < numOfTxs {
+		rndInt := rand.Uint64()
+		idx := rndInt % uint64(spaceSize)
+		idxs[idx] = struct{}{}
+	}
+	return idxs
+}
+
+func (t *TxMempool) Put(id types.TransactionId, tx *types.Transaction) {
+	t.mu.Lock()
+	t.txs[id] = tx
+	t.getOrCreate(tx.Origin()).Add(0, tx)
+	t.mu.Unlock()
+}
+
+func (t *TxMempool) Invalidate(id types.TransactionId) {
+	t.mu.Lock()
+	if tx, found := t.txs[id]; found {
+		if pendingTxs, found := t.accounts[tx.Origin()]; found {
+			pendingTxs.RemoveNonce(tx.AccountNonce, func(id types.TransactionId) {
+				delete(t.txs, id)
+			})
+			if pendingTxs.IsEmpty() {
+				delete(t.accounts, tx.Origin())
+			}
+		}
```

Comment:
consider printing a fat error in the `else` case

Reply:
Do you find it erroneous that we encounter a tx in a block that we don't have in the mempool?

I mean, it might not be the common case, and we may want to log it, but **fat** error? IDK

Reply:
Since the mempool doesn't currently have a logger, I wonder if it's worth the effort